bwe/BandwidthEstimator.cpp

```diff
@@ -31,164 +43,243 @@ BandwidthEstimator::BandwidthEstimator(const Config& config)
     }
-    statePredictionCovariance *= _weightMeanCovariance;
+    statePredictionCovariance *= _weightCovariance;
     statePredictionCovariance += (_weightCovariance0 * math::outerProduct(sigmaPoints[0]));
     assert(math::isValid(statePredictionCovariance));
 
-    double predictedMeanDelay = predictedDelays[0] * _weightMean0;
-    for (size_t i = 1; i < predictedDelays.size(); ++i)
-    {
-        predictedMeanDelay += predictedDelays[i] * _weightMeanCovariance;
-    }
-
     const auto residual0 = predictedDelays[0] - predictedMeanDelay;
     double covDelay = _weightCovariance0 * residual0 * residual0;
     math::Matrix<double, 3> crossCovariance;
     for (size_t i = 1; i < predictedDelays.size(); ++i)
     {
         const auto residual = predictedDelays[i] - predictedMeanDelay;
-        covDelay += _weightMeanCovariance * residual * residual;
+        covDelay += _weightCovariance * residual * residual;
         crossCovariance += residual * sigmaPoints[i];
     }
-    crossCovariance *= _weightMeanCovariance;
+    crossCovariance *= _weightCovariance;
     crossCovariance += _weightCovariance0 * residual0 * sigmaPoints[0];
 
+    // update position towards mean position
     const auto kalmanGain = crossCovariance * (1.0 / covDelay);
     _state = predictedMeanState + kalmanGain * (observedDelay - predictedMeanDelay);
-    _covarianceP = statePredictionCovariance - kalmanGain * covDelay * math::transpose(kalmanGain);
+    _covarianceP = statePredictionCovariance - crossCovariance * math::transpose(kalmanGain);
+
+    sanitizeState(observedDelay, packetSize * 8, _state);
+
+    assert(math::isValid(_covarianceP));
     math::makeSymmetric(_covarianceP);
     assert(math::isSymmetric(_covarianceP));
     assert(math::isValid(_covarianceP));
 
     _observedDelay = observedDelay;
-    _state(0) = std::max(0.0, std::min(_state(0), _config.maxNetworkQueue * 8));
 
-    if (_state(2) > observedDelay + 12.0)
+    _previousReceiveTime = receiveTimeNs;
+    _previousTransmitTime = transmitTimeNs;
+}
+
+void BandwidthEstimator::sanitizeState(const double observedDelay,
+    const double packetBits,
+    math::Matrix<double, 3>& state)
+{
+    state(Bandwidth) = math::clamp(state(Bandwidth), _config.modelMinBandwidth, _config.estimate.maxKbps);
+
+    state(QueuedBits) = std::max(packetBits, state(QueuedBits));
+
+    if (observedDelay - predictAbsoluteDelay(state) < 0 && state(QueuedBits) > _config.mtu * 3)
     {
-        // under estimating delay. Likely in congestion and over estimating bw
-        _congestion.margin = _config.congestion.backOff;
-        _state(1) = std::min(_state(1), _receiveBandwidth.get(550 * utils::Time::ms) * utils::Time::ms);
+        // adjust queue if we received packet earlier than expected. If bw is higher than expected, queue should also
+        // have drained more
+        const double delayErr = predictAbsoluteDelay(state) - observedDelay;
```

Comment:
This avoids bw over estimation when bw goes up. The model queue is always lagging behind as it is modeled on previous lower drain rate.